include/graphblas/bsp/collectives_blas1.hpp

```diff
@@ -91,28 +96,35 @@ namespace grb {
 		RC gather(
 			const IOType &in,
 #ifdef BLAS1_RAW
```

Comment:
This ubiquitous `#ifdef BLAS1_RAW ... #else ....` makes the code harder to read and to maintain; also, in case of transfer to a dedicated library (like a library of LPF primitives or LPF itself), the person in charge of the transfer should "disentangle" the code, because `grb` types like `Coordinates` or `Vector` are internal to this project and not exported (can you also foresee the "entangling" of header dependencies, or the pain to "disantangle"?).

The proposed solution is to split the `gather` into two dedicated versions, one working with only C-style arrays and one with `grb` types only, using overloading. For example

```c++
// function with I/O logic
template<
    descr = descriptors::no_operation,
	typename IOType
>
RC gather(
    IOType &in,
	IOType * const out,
    const lpf_pid_t root
) {
 <gather functionality here>
}

// function with only "translation" logic
template<
    Descriptor descr = descriptors::no_operation,
	typename IOType,
	typename Coords
>
RC gather(
	const IOType &in,
	Vector< IOType, reference, Coords > &out,
	const lpf_pid_t root
		) {
    // overlap/nullity checks, or whatever
    RC ret = gather<descr, IOType>( in, internal::getRaw( out ), root );
    if( ret == SUCCESS && internal::grb_BSP1D.load().s == root ) {
        // make sure sparsity info is correct
        internal::getCoordinates( out ).template assignAll< true >();
    }
    return ret;
}
```

There is also another possible step. The second function needs the definition of `grb::Vector`, which is ideally something higher level than these functionalities: a method of `grb::Vector` may theoretically call `gather`. In such a case, we would have circular dependencies: `gather` needs the definition of `grb::Vector`, which needs the definition of `gather`. You may already foresee the pain, solved via careful `#include` order engineering (better: crafting) or forward definitions. Therefore, I would suggest to place the second function (with only "translation" logic) into a higher-level header, like `bsp/vector.hpp` itself or the likes.